### PR TITLE
[pipeline] add pipeline support for all T5 models

### LINKED ISSUE #4247: [shardformer] support pipeline for t5
support pipeline for T5 models, integrate its policy to shardformer, and add relevant tests.

### PR DESCRIPTION
**Pull request by Fridge003**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

closed #4247

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

- rewrite policies and pipeline forward methods for T5Model and T5ForConditionalGeneration

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by Fridge003** [comment]
截屏2023-07-24 15 05 29

File: colossalai/shardformer/modeling/t5.py [hunk 1]
```diff
@@ -238,6 +242,299 @@ def custom_forward(*inputs):
...
+                output_attentions=output_attentions,
+                output_hidden_states=output_hidden_states,
+                return_dict=return_dict,
+                stage_manager=stage_manager,
+                hidden_states=hidden_states,
+                position_bias=position_bias,
+                encoder_decoder_position_bias=encoder_decoder_position_bias,
+                stage_index=stage_index,
+                decoder_starting_stage=decoder_starting_stage)
+            return encoder_outputs
```
[the remark this answers is not in the document]
Comment (Fridge003) [reply colossalai/shardformer/modeling/t5.py:325]: I will raise ValueError if in decoder and encoder_ouputs is None, please refer to line 328-329. It should be scheduler's responsibility to provide non-empty encoder_outputs for decoder stages.

[the remark this answers is not in the document]
Comment (Fridge003) [reply colossalai/shardformer/modeling/t5.py:325]: The algorithm of t5 layer distribution ensures that encoder and decoder must reside in different stages. So the case when encoder and decoder are in the same stage will not happen.

**Review by CjhHa1 — APPROVED** [review]
LGTM

**Pull request merged**